### ISSUE TITLE
[GSC] Unable to converting docker image running python flask

### ISSUE THREAD

**Issue by ADL-work** [oscarlab/graphene#2261]
## Description of the problem
I am trying to converting docker image running a simple flask server using GSC, but failed to run the converted docker image.
Given that I have all needed SGX PSW and SGX SDK installed (I managed to run the examples in graphene/Examples/ folder)

## Steps to reproduce

testflask.py
```
#! /usr/bin/python3

from flask import Flask
app = Flask(__name__)

@app.route('/')
def hello_world():
    return 'Hey, we have Flask in a Docker container!'


if __name__ == '__main__':
    app.run(debug=True, host='0.0.0.0', port=5000)
```

testflask.dockerfile
```
From ubuntu:18.04

RUN apt-get update \
    && apt-get install -y python3 python3-pip \
    && pip3 install flask \
    && mkdir -p /graphene/Examples

# Build environment of this Dockerfile should point to the root of Graphene's Examples/
COPY testflask.py /graphene/Examples

WORKDIR /graphene/Examples

CMD ["python3"]
```

testflask.yaml
```
Distro: "ubuntu18.04"
Graphene:
    Repository: "https://github.com/oscarlab/graphene.git"
    Branch:     "master"
SGXDriver:
    Repository: "https://github.com/01org/linux-sgx-driver.git"
    Branch:     "sgx_driver_1.9"
```

Following by conversion commands 

Build the unsigned image
`./gsc build --insecure-args -d testflask -c ./test/testflask.yaml ./test/testflask.manifest`

Sign the image
`./gsc sign-image -c ./test/testflask.yaml testflask ./enclave-key.pem`

## Expected results
After running the command
`docker run -p 5000:5000 -it --device /dev/isgx:/dev/isgx --device /dev/gsgx:/dev/gsgx -v /var/run/aesmd/aesm.socket:/var/run/aesmd/aesm.socket  gsc-testflask`

The Flask application will be up and running for the port 5000 and the request returns a message
```
curl localhost:5000 
```
`Hey, we have Flask in a Docker container!`

## Actual results

Error
`curl: (56) Recv failure: Connection reset by peer`

When login into the docker container, I can actually start the Flask server by running 
`python3 testflask.py`

Which mean the `docker run` command for the converted image didn't start the server

Do you have any suggestions so that I can start the web server with GSC?
Plus, I see from the Examples that we can also run the GSC image with arguments, I also tried but the container stops after finalizing the SGX environment, do you have any other examples so that I can refer too?

Thank you a lot for the help!

**Comment by dimakuv** [comment]
> After running the command
`docker run -p 5000:5000 -it --device /dev/isgx:/dev/isgx --device /dev/gsgx:/dev/gsgx -v /var/run/aesmd/aesm.socket:/var/run/aesmd/aesm.socket gsc-testflask`

This command simply runs the Python interpreter because you didn't specify any arguments on the command line, and your `CMD` in the Dockerfile also doesn't have any arguments.

You should specify the arguments. Something like:
```
docker run -p 5000:5000 -it --device /dev/isgx:/dev/isgx --device /dev/gsgx:/dev/gsgx -v /var/run/aesmd/aesm.socket:/var/run/aesmd/aesm.socket gsc-testflask /graphene/Examples/testflask.py
```

Note that because you specified `--insecure-args` during GSC build, you can use command-line arguments without problems.

For more context, see the examples and pay attention to the arguments at the end of `docker run ...` command:
- https://graphene.readthedocs.io/en/latest/manpages/gsc.html#example
- https://github.com/oscarlab/graphene/blob/master/Tools/gsc/test/Makefile#L119
- https://github.com/oscarlab/graphene/blob/master/Tools/gsc/test/Makefile#L125

**Comment by ADL-work** [comment]
Hi @dimakuv, thank you for pointing that out,

Yes, I have run the suggested command:

`docker run -d -p 5000:5000 --device /dev/isgx:/dev/isgx --device /dev/gsgx:/dev/gsgx -v /var/run/aesmd/aesm.socket:/var/run/aesmd/aesm.socket  gsc-testflask /graphene/Examples/testflask.py`

I still have the error

`curl: (56) Recv failure: Connection reset by peer`

and when I went into the container, I can manually start the Flask server there.

Additionally, I tried another setting in the `testflask.dockerfile`:

`CMD ["python3", "testflask.py"]`

so that the container will start the Flask server after it is initiated. The `docker run` command will become

`docker run -d -p 5000:5000 --device /dev/isgx:/dev/isgx --device /dev/gsgx:/dev/gsgx -v /var/run/aesmd/aesm.socket:/var/run/aesmd/aesm.socket  gsc-testflask`

But the container behaves the same like the previous scenario: cannot reach the server, but I could get in the docker container & manually start the server.

I followed the example here
- https://github.com/oscarlab/graphene/blob/master/Tools/gsc/test/Makefile#L149

And I can actually start the server, so I think there might be something with setting up Flask?

Thank you for your help!

**Comment by dimakuv** [comment]
I see, so you already did all these steps. Looks like something is missing to run Flask correctly in Graphene.

This could be the same problem as in https://github.com/oscarlab/graphene/issues/1767.

I created a PR #2204 some time ago to fix this. Could you check it? Or wait until it is merged.

Can you also show your `test/testflask.manifest` contents? Maybe something is wrong there.

Also, you seem to already build Graphene with debugging enabled. Do you see the debug log of Graphene when you do `docker run`? Do you notice anything interesting in this log? Maybe Graphene complains about something?

**Comment by ADL-work** [comment]
Hi, thanks for the hint, I will take a look in the #1767

Oh, I forgot to share the `testflask.manifest` file, basically I just replicated from some existing samples:

```
sgx.enclave_size = "2G"
sgx.thread_num = 16
```
Might be the manifest does not list all the required files/settings to be used.

Yes, I could see Graphene debug logs printing during the run, mostly relating to `shim` activities for python3 (`[P8:python3]` ...) but nothing particular from what I observed.

**Comment by ADL-work** [comment]
Hi @dimakuv , 

Do you have any updates on your side?
I've tried to run with different commit versions, unfortunately non of it works so far.
Thank you for the help!

**Comment by dimakuv** [comment]
No, I haven't looked at this.

> Also, you seem to already build Graphene with debugging enabled. Do you see the debug log of Graphene when you do docker run? Do you notice anything interesting in this log? Maybe Graphene complains about something?

Could you do this? Could you look at the Graphene log? Or at least attach the Graphene debug log here.

**Comment by ADL-work** [comment]
Hi @dimakuv 

Here is the log from the docker run output of the converted image
[debug.log](https://github.com/oscarlab/graphene/files/6318872/debug.log)

There was an error at the end mentioning 
`FileNotFoundError: [Errno 2] No such file or directory: '/entrypoint': '/entrypoint'`

My dockerfile is below, I name it as `testflask1.dockerfile`
```
From ubuntu:18.04

RUN apt-get update \
    && apt-get install -y python3 python3-pip\
    && mkdir -p /graphene/Examples \
    && pip3 install --upgrade --no-cache-dir pip \
    && pip3 install flask 

RUN pip3 install --no-cache-dir flask

# Build environment of this Dockerfile should point to the root of Graphene's Examples/
COPY testflask.py /graphene/Examples

RUN chmod 777 /graphene/Examples/testflask.py

WORKDIR /graphene/Examples

ENTRYPOINT [ "python3" ]

CMD [ "testflask.py" ]
```

`testflask.yaml` file
```
Distro: "ubuntu18.04"
Graphene:
    Repository: "https://github.com/oscarlab/graphene.git"
    Branch:     "master"
SGXDriver:
    Repository: "https://github.com/01org/linux-sgx-driver.git"
    Branch:     "sgx_driver_1.9"

```

`testflask1.manifest` file
```
sgx.allow_file_creation = 1
sgx.enclave_size = "2G"
sgx.thread_num = 128
# sgx.file_check_policy = "allow_all_but_log"
# sgx.ra_accept_group_out_of_date = 1

sys.stack.size = "256K"
sys.brk.size = "4M"
sgx.allowed_files.etchostname = "file:/etc/hostname"
sgx.allowed_files.etchosts    = "file:/etc/hosts"
sgx.allowed_files.etcresolv   = "file:/etc/resolv.conf"


sgx.allowed_files.testflask = "file:/graphene/Examples/testflask.py"
sgx.trusted_files.testflask = "file:/graphene/Examples/testflask.py"
```

And the command to run after converting:
`docker run -it -p 5000:5000 --device /dev/isgx:/dev/isgx --device /dev/gsgx:/dev/gsgx -v /var/run/aesmd/aesm.socket:/var/run/aesmd/aesm.socket  gsc-testflask1`

**Comment by dimakuv** [comment]
Thanks, this is much better with the Graphene log.

- You have a bunch of `error: Accessing file:/tmp/ccnLCaRy.le is denied (Operation denied)` errors. This makes sense because these temporary files are created on the fly by different Graphene processes, and they are not allowed in your manifest. For testing, please add `sgx.allowed_files.tmpfiles = "file:/tmp"` to your `testflask1.manifest` file.

- The real problem in your Flask app is that it wants to do the following: ` * Restarting with stat`. And then it proceeds with `trace: ---- shim_execve("/entrypoint", [/entrypoint,/graphene/Examples/testflask.py,], ...`. Of course, `entrypoint` is not a real executable, there must be a Python3 executable in its place. Please add `loader.argv0_override = "python3"` to your `testflask1.manifest` file.

So, to summarize, adding the below two lines to your manifest file should unblock you:
```
sgx.allowed_files.tmpfiles = "file:/tmp"
loader.argv0_override = "python3"
```

**Comment by dimakuv** [comment]
Also, you have some (non-critical) issues in your manifest file:
```
# sgx.ra_accept_group_out_of_date = 1                 -- this doesn't exist!
loader.env.RA_TLS_ALLOW_OUTDATED_TCB_INSECURE = "1"   -- this is the correct way
```
```
# both of the below are not needed, just remove these lines (why did you add them in the first place?)
sgx.allowed_files.testflask = "file:/graphene/Examples/testflask.py"
sgx.trusted_files.testflask = "file:/graphene/Examples/testflask.py"
```

**Comment by dimakuv** [comment]
The bug in GSC that allowed this to happen is this: https://github.com/oscarlab/graphene/blob/master/Tools/gsc/templates/entrypoint.manifest.template#L21.

As you can see, GSC template misspells `loader.arg0_override` (it must be `loader.argv0_override`, with a `v`!). I will submit a quick PR to fix it.

**Comment by ADL-work** [comment]
Hi @dimakuv 
Yes, some of the issues from the manifest file are basically my trial-and-error attempts to test the container, thanks for clarifying them out :)

Thanks to your suggestion, now I have this manifest file for `testflask1.manifest`

```
sgx.enclave_size = "4G"
sgx.thread_num = 128

sys.stack.size = "256K"
sys.brk.size = "4M"
sgx.allowed_files.etchostname = "file:/etc/hostname"
sgx.allowed_files.etchosts    = "file:/etc/hosts"
sgx.allowed_files.etcresolv   = "file:/etc/resolv.conf"

sgx.allowed_files.tmpfiles = "file:/tmp"
sgx.allowed_files.usrfiles = "file:/usr"
loader.argv0_override = "python3"
```

I added the `/usr` folder because during the run I encountered the error `error: Accessing file: ...` with `/usr` and `/usr/lib` or `/usr/lib/python3.6`.

I ran the conversion and the signed docker container, but the `docker run` was killed, even though the container was still running in the background, here is the log [testflask1.log](https://github.com/oscarlab/graphene/files/6323543/testflask1.log)

I see the errors:
```
error: Failed to exchange the secret key via RPC: -24
error: Failed to initialize child process: -24
```

and still many about the files
```
error: Accessing file:/usr/lib/python3.6/encodings/__pycache__/__init__.cpython-36.pyc is denied (Operation denied). This file is not trusted or allowed. Trusted files should be regular files (seekable).
```

**Comment by dimakuv** [comment]
You should **not** add the `.pyc` files to the manifest file. These files are Python's "pre-compiled" versions of `.py` files, and Python doesn't really need them. So these errors are expected. (Also, my person opinion, but Python runtime is absolutely terrible, and someone should rewrite Python's internals to be adequate.)

I am currently looking at your log to see what goes wrong.

**Comment by dimakuv** [comment]
At this point, I can't help much. You correctly identified the root cause of the failure, it is these lines:
```
debug: enclave initializing:
debug:     enclave id:   0x00000000fffff000
debug:     mr_enclave:   0be660bd23d79930b995b7d46a1d64d63518dffb8f77d5e56aeb89734a2e4bc0
error: Failed to exchange the secret key via RPC: -24
error: Failed to initialize child process: -24
[P94:i1:python3] debug: Port 0xf86fc150 (handle 0xf963b9f0) disconnected
[P94:i1:python3] debug: Deleting port 0xf86fc150 (handle 0xf963b9f0) of process 97
[P94:i1:python3] debug: Child process (vmid: 0x61) got disconnected
[P94:T165:python3] trace: ---- return from shim_wait4(...) = -512
```

So basically the parent enclave created (yet another) child enclave, but the child enclave didn't perform a handshake with the parent enclave. And the parent enclave said "My child enclave unexpectedly died, I don't know how to handle this so I'll die myself" (well, kinda, the parent seems to await a child enclave to establish a TCP connection to it, but doesn't find this connection and then dies).

It looks like your Linux decided to kill the Docker container process? There is a well-known limitation of SGX that there cannot be more total enclave memory (used by all running enclaves) than the total amount of RAM + swap file. How much RAM do you have on your system? If it's e.g. 8-16GB, I can imagine Linux deciding that your Flask-SGX workload consumes too much RAM (recall that SGX enclave pages actually reside in RAM, just encrypted), and Linux just kills the whole thing.

**Comment by dimakuv** [comment]
Ok, to test this hypothesis of Linux killing your Docker container because there is too few RAM left on the system, you can change your manifest file:
```
# sgx.enclave_size = "4G"  --  that's too much enclave memory, Linux may kill your enclaves!
sgx.enclave_size = "1G"
```

Also, why do you need some many threads (`sgx.thread_num = 128`)? Just specify a small number, e.g., `8` or `16`. Python is single-threaded anyway, right? Or does Flask use some multi-threading library underneath?

**Comment by ADL-work** [comment]
Hi @dimakuv  thanks for pointing it out.

I have modified `sgx.enclave_size = "1G"` and `sgx.thread_num = 16`

Now the docker container isn't killed any more, the docker app keeps running, but the `curl` request is hanged without any response.

I am currently using an Azure Confidential Computing VM with Standard DC4s_v2 (4 vcpus, 16 GiB memory)
Here is the latest log file 
[testflask1_4.log](https://github.com/oscarlab/graphene/files/6326086/testflask1_4.log)

I let the docker app run for 22 mins but it seems the Flask server didn't bring up after the `Restarting with stats`
```
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                           PORTS                    NAMES
58a93bb3129d        gsc-testflask1      "/bin/bash /apploade…"   22 minutes ago      Exited (130) 12 minutes ago                               hardcore_dirac
```

Also I experienced a bit lag in the VM after running some conversions and docker apps. It also stopped the VM when I tried to run the `docker run` for the converted images. DId you experience something similar before?

Thank you for the help!

**Comment by dimakuv** [comment]
Unfortunately, I don't know how to proceed with this issue. It will take too much effort to reproduce (install Flask-on-Graphene in Azure VM), and there is no obvious error visible here.

@ADL-work Maybe you already figured it out? Is this still relevant?

**Comment by ADL-work** [comment]
@dimakuv unfortunately I still have the issue. I will try a couple more with different settings, if it doesn't work, I will install Flask-on-Graphene and run from there.
Thanks for your help!

**Comment by mkow** [comment]
I'm closing this, as we don't have enough information to either analyze the issue of easily reproduce ourselves.

@ADL-work we can reopen it if you provide the above.
